workflows/plan_manager.py

```diff
@@ -0,0 +1,731 @@
+    content: str
+    step_progress_counter: int
+    state: Literal["pending", "completed", "failed", "in_progress"]
+    reflection: Optional[str]
+    messages: List[BaseAgentEvent | BaseChatMessage]
+
+    def dump(self) -> Dict:
+        """
+        Serialize the Step object into a dictionary.
+
+        Returns:
+            Dict: A dictionary containing the step's attributes, with messages serialized as a list.
+        """
+        return {
+            "step_id": self.step_id,
+            "content": self.content,
+            "step_progress_counter": self.step_progress_counter,
+            "state": self.state,
+            "reflection": self.reflection,
+            "messages": [msg.dump() for msg in self.messages] if self.messages else [],
+            # Messages sent by the Orchestrator agent and tool agents
+        }
+
+    @classmethod
+    def load(cls, data: Dict) -> "Step":
+        """
+        Deserialize a dictionary into a Step object.
+
+        Args:
+            data (Dict): The dictionary containing step data.
+
+        Returns:
+            Step: A new Step instance populated with the provided data.
+        """
+        return cls(
+            step_id=data["step_id"],
+            content=data["content"],
+            step_progress_counter=data.get("step_progress_counter", 0),
+            state=data.get("state", "pending"),
+            reflection=data.get("reflection"),
+            messages=[BaseMessage.load(msg) for msg in data.get("messages", [])],
+        )
+
+
+class Plan(BaseModel):
+    """
+    Represents a plan consisting of multiple steps, identified by a unique ID.
+
+    Attributes:
+        plan_id (str): Unique identifier for the plan.
+        task (str): Description of the task this plan addresses.
+        steps (OrderedDict[str, Step]): Dictionary mapping step IDs to Step objects.
+        awaiting_confirmation (bool): Whether the plan awaits user confirmation.
+        summary (Optional[str]): Summary of the plan.
+    """
+
+    plan_id: str
+    task: str
+    steps: OrderedDict[str, Step]
+    awaiting_confirmation: bool
+    summary: Optional[str]
+
+    def get_current_step(self) -> Optional[Tuple[str, str]]:
+        """
+        Retrieve the first step that is either pending or in progress.
+        Return the step_id and step content.
+
+        Returns:
+            Optional[Tuple[str, str]]: A tuple of (step_id, content) if found, else None.
+        """
+        for step_id, step in self.steps.items():
+            if step.state in ["pending", "in_progress"]:
+                return (step_id, step.content)
+        return None
+
+    def get_all_contents(self) -> List[str]:
+        """
+        Retrieve the content of all steps in the plan.
+
+        Returns:
+            List[str]: A list containing the content of each step in the plan.
+        """
+        return [step.content for step in self.steps.values()]
+
+    def get_all_states(self) -> "OrderedDict[str, str]":
```

Comment:
Redundant double quotes ("OrderedDict[str, str]") here,
```suggestion
    def get_all_states(self) -> OrderedDict[str, str]:
```